lib/Conversion/HandshakeToDC/HandshakeToDC.cpp

```diff
@@ -0,0 +1,612 @@
+    auto selectToken = select.token;
+    bool isIndexType = selectData.getType().isa<IndexType>();
+
+    bool withData = !op.getResult().getType().isa<NoneType>();
+
+    llvm::SmallVector<DCTuple> inputs;
+    for (auto input : adaptor.getDataOperands())
+      inputs.push_back(unpack(rewriter, input));
+
+    Value dataMux;
+    Value controlMux = inputs.front().token;
+    // Convert the data-side mux to a sequence of arith.select operations.
+    // The data and control muxes are assumed one-hot and the base-case is set
+    // as the first input.
+    if (withData)
+      dataMux = inputs[0].data.front();
+
+    llvm::SmallVector<Value> controlMuxInputs = {inputs.front().token};
+    for (auto [i, input] :
+         llvm::enumerate(llvm::make_range(inputs.begin() + 1, inputs.end()))) {
+      if (withData) {
+        Value cmpIndex;
+        Value inputData = input.data.front();
+        Value inputControl = input.token;
+        if (isIndexType) {
+          cmpIndex = rewriter.create<arith::ConstantIndexOp>(op.getLoc(), i);
+        } else {
+          size_t width = selectData.getType().cast<IntegerType>().getWidth();
+          cmpIndex =
+              rewriter.create<arith::ConstantIntOp>(op.getLoc(), i, width);
+        }
+        auto inputSelected = rewriter.create<arith::CmpIOp>(
+            op.getLoc(), arith::CmpIPredicate::eq, selectData, cmpIndex);
+        dataMux = rewriter.create<arith::SelectOp>(op.getLoc(), inputSelected,
+                                                   inputData, dataMux);
+
+        // Legalize the newly created operations.
+        convertedOps->insert(cmpIndex.getDefiningOp());
+        convertedOps->insert(dataMux.getDefiningOp());
+        convertedOps->insert(inputSelected);
+
+        // And similarly for the control mux, by muxing the input token with a
+        // select value that has it's control from the original select token +
+        // the inputSelected value.
+        auto inputSelectedControl =
+            pack(rewriter, selectToken, ValueRange{inputSelected});
+        controlMux = rewriter.create<dc::SelectOp>(
+            op.getLoc(), inputSelectedControl, inputControl, controlMux);
+        convertedOps->insert(controlMux.getDefiningOp());
+      }
+    }
+
+    // finally, pack the control and data side muxes into the output value.
+    rewriter.replaceOp(op, pack(rewriter, controlMux,
+                                withData ? ValueRange{dataMux} : ValueRange{}));
+    return success();
+  }
+};
+
+static hw::ModulePortInfo getModulePortInfo(TypeConverter &tc,
+                                            handshake::FuncOp funcOp) {
+  hw::ModulePortInfo ports({}, {});
+  auto *ctx = funcOp->getContext();
+  auto ft = funcOp.getFunctionType();
+
+  // Add all inputs of funcOp.
+  unsigned inIdx = 0;
+  for (auto [index, type] : llvm::enumerate(ft.getInputs())) {
+    ports.inputs.push_back({StringAttr::get(ctx, "in" + std::to_string(index)),
```

Comment:
Can you use `emplace_back` instead, or does this only work when there is an implemented constructor?